core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java

```diff
@@ -0,0 +1,257 @@
+      Collection<PageInfo> pageInfos = pageStore.getPages();
+      for (PageInfo pageInfo : pageInfos) {
+        PageId pageId = pageInfo.getPageId();
+        metaStore.addPage(pageId, pageInfo);
+        evictor.updateOnPut(pageId);
+      }
+      return new LocalCacheManager(conf, metaStore, pageStore, evictor);
+    } catch (Exception e) {
+      try {
+        pageStore.close();
+      } catch (Exception ex) {
+        e.addSuppressed(ex);
+      }
+      throw new IOException("failed to create local cache manager", e);
+    }
+  }
+
+  /**
+   * @param conf the Alluxio configuration
+   * @param evictor the eviction strategy to use
+   * @param metaStore the meta store manages the metadata
+   * @param pageStore the page store manages the cache data
+   */
+  @VisibleForTesting
+  LocalCacheManager(AlluxioConfiguration conf, MetaStore metaStore,
+      PageStore pageStore, CacheEvictor evictor) {
+    mMetaStore = metaStore;
+    mPageStore = pageStore;
+    mEvictor = evictor;
+    mPageSize = conf.getBytes(PropertyKey.USER_CLIENT_CACHE_PAGE_SIZE);
+    mCacheSize = (long) (conf.getBytes(PropertyKey.USER_CLIENT_CACHE_SIZE)
+        / (1.0 + pageStore.getOverheadRatio()));
+    for (int i = 0; i < LOCK_SIZE; i++) {
+      mPageLocks[i] = new ReentrantReadWriteLock();
+    }
+  }
+
+  /**
+   * Gets the lock for a particular page. Note that multiple pages may share the same lock as lock
+   * striping is used to reduce resource overhead for locks.
+   *
+   * @param pageId page identifier
+   * @return the corresponding page lock
+   */
+  private ReadWriteLock getPageLock(PageId pageId) {
+    return mPageLocks
+        [Math.floorMod((int) (pageId.getFileId().hashCode() + pageId.getPageIndex()), LOCK_SIZE)];
+  }
+
+  /**
+   * Gets a pair of locks to operate two given pages. One MUST acquire the first lock followed by
+   * the second lock.
+   *
+   * @param pageId page identifier
+   * @param pageId2 page identifier
+   * @return the corresponding page lock pair
+   */
+  private Pair<ReadWriteLock, ReadWriteLock> getPageLockPair(PageId pageId, PageId pageId2) {
+    if (pageId.getFileId().hashCode() + pageId.getPageIndex()
+        < pageId2.getFileId().hashCode() + pageId2.getPageIndex()) {
+      return new Pair<>(getPageLock(pageId), getPageLock(pageId2));
+    } else {
+      return new Pair<>(getPageLock(pageId2), getPageLock(pageId));
+    }
+  }
+
+  @Override
+  public boolean put(PageId pageId, byte[] page) throws IOException {
+    PageId victim = null;
+    PageInfo victimPageInfo = null;
+    boolean enoughSpace;
+
+    ReadWriteLock pageLock = getPageLock(pageId);
+    try (LockResource r = new LockResource(pageLock.writeLock())) {
+      try (LockResource r2 = new LockResource(mMetaLock.writeLock())) {
+        if (mMetaStore.hasPage(pageId)) {
+          LOG.debug("{} is already inserted before", pageId);
+          return false;
+        }
+        enoughSpace = mPageStore.bytes() + page.length <= mCacheSize;
```

Comment:
@bf8086 From what I understand, `mPageStore.bytes()` is not protected by `mMetaLock`, so if `enoughSpace` is true here, there is no guarantee that it will continue to hold true.

Reply:
Instead of `mPageStore.bytes()` I think we need to use something like `mMetastore.bytes()`? Since we can guarantee that is not changing and we will reserve the required space while still holding the metalock